stix2/base.py

```diff
@@ -347,6 +360,61 @@ def _check_property(self, prop_name, prop, kwargs):
             for ref in kwargs[prop_name]:
                 self._check_ref(ref, prop, prop_name)
 
+    def _generate_id(self, kwargs):
+        required_prefix = self._type + "--"
+        namespace = uuid.UUID("00abedb4-aa42-466c-9c01-fed23315a9b7")
+
+        try:
+            properties_to_use = self._id_contributing_properties
+            if properties_to_use:
+                streamlined_object = {}
+                if "hashes" in kwargs and "hashes" in properties_to_use:
+                    possible_hash = self._choose_one_hash(kwargs["hashes"])
+                    if possible_hash:
+                        streamlined_object["hashes"] = possible_hash
+                for key in kwargs.keys():
+                    if key in properties_to_use and key != "hashes":
+                        if type(kwargs[key]) is dict:
+                            for otherKey in kwargs[key]:
+                                if isinstance(kwargs[key][otherKey], _STIXBase):
+                                    streamlined_object[key] = self._embed_obj_to_json(kwargs[key][otherKey])
+                                else:
+                                    streamlined_object[key] = kwargs[key]
+                        else:
+                            if isinstance(kwargs[key], _STIXBase):
+                                streamlined_object[key] = self._embed_obj_to_json(kwargs[key])
+                            else:
+                                streamlined_object[key] = kwargs[key]
+
+                if streamlined_object:
+                    data = canonicalize(str(streamlined_object), utf8=False)
```

Comment:
If I understand what's going on here, I don't think this is what you meant.  str(...) produces a string.  All this code is going to do is probably escape some junk in the string to make it JSON-compliant.  I think what you meant to do is run their encoder over the whole python structure (of streamlined_object), to produce canonical JSON.

Reply:
Also, it looks like you made a dict to be encoded as (canonical) JSON.  But the STIX spec says:

> The value of the name portion SHOULD be the list of "ID Contributing Properties" ...

Which probably means a *list* of ID Contributing Property *values*.  Not an object.  Does that seem right?  Or did we decide to go against the "should"? :)

Reply:
The way it's written in the spec seems like it could be interpreted in different ways... I'm okay with either changing it to a list or leaving it a dict until we get more clarification in the spec.